### PR TITLE
Rework build & packaging duration measurements

### PR DESCRIPTION
**Pull request by chouquette**
This provides a way to use duration computed by `measure` and generate a report with those, but without breaking failure detection this time

### CODE REVIEW

**Comment by chouquette** [comment]
test build: https://gitlab.ddbuild.io/DataDog/datadog-agent/-/pipelines/29385555

File: lib/omnibus/builder.rb [hunk 1]
```diff
@@ -762,12 +762,11 @@ def build
         end
       end
 
-      build_time = measure("Build #{software.name}") do
+      measure("Build #{software.name}", lambda { |d| software.store_build_duration(d) }) do
```
Comment (alopezz) [inline lib/omnibus/builder.rb]: I think this should work, too:

```suggestion
      measure("Build #{software.name}", software.method(:store_build_duration) do
```

File: lib/omnibus/packagers/base.rb [hunk 2]
```diff
@@ -177,11 +177,11 @@ def skip_packager(val = false)
     #   - setup
     #   - build
     #
-    def run!
+    def run!(duration_cb = nil)
```
Comment (alopezz) [inline lib/omnibus/packagers/base.rb]:
```suggestion
    def run!
```
I'd rather avoid `run!` from having this sort of side-function; see next comment for idea.

File: lib/omnibus/packagers/base.rb [hunk 3]
```diff
@@ -177,11 +177,11 @@ def skip_packager(val = false)
...
       # Ensure the package directory exists
       create_directory(Config.package_dir)
 
-      measure("Packaging time") do
+      measure("Packaging time", duration_cb) do
```
Comment (alopezz) [inline lib/omnibus/packagers/base.rb]:
```suggestion
      measure("Packaging time", ->(duration) { project.package_summary[id] = duration }) do
```

That would require making `project.package_summary` public or adding some sort of explicit setter method such as `project.record_duration(:packaging, id, duration)`.

Similarly for the `stripper`.

File: lib/omnibus/instrumentation.rb [hunk 4]
```diff
@@ -18,12 +18,15 @@ module Omnibus
...
+    def measure(label, duration_cb = nil)
       start = Time.now
       yield
     ensure
       elapsed = Time.now - start
       log.info(log_key) { "#{label}: #{elapsed.to_f.round(4)}s" }
+      if duration_cb
+        duration_cb.call(elapsed)
+      end
       elapsed
```
Comment (alopezz) [inline lib/omnibus/instrumentation.rb]:
```suggestion
```

**Review by alopezz — COMMENTED** [review]
I hadn't foreseen the other places that were affected by the measuring logic, and I'd rather not have to thread the callback parameter through intermediate methods.

I have a possible suggestion for that if we keep the callback approach, but consider also:

- We could make `measure` simply return the elapsed time as originally intended by adding a single line (arguably a much easier and simpler change). However I find it a bit strange to return durations from `run!` methods (plus at least in the stripper it would change the documented return value).
- We could consider moving the `measure` calls inside the `run!` methods out to the caller, which could be a little bit more cleaner; I don't know if there are other implications though.

File: lib/omnibus/packagers/base.rb (hunk 3 above)
Comment (chouquette) [reply lib/omnibus/packagers/base.rb]: I think I understood your proposal and I like it better that the current one. I'll attempt to implement it asap
Thanks for the idea!

**Review by alopezz — APPROVED** [review]
Looks good 👍 . Probably worth another test run before merging though.

**Comment by chouquette** [comment]
Agreed :) https://gitlab.ddbuild.io/DataDog/datadog-agent/-/pipelines/29402777

**Pull request merged**